source/extensions/filters/network/thrift_proxy/router/router_impl.cc

```diff
@@ -263,6 +262,7 @@ FilterStatus Router::messageBegin(MessageMetadataSharedPtr metadata) {
   Tcp::ConnectionPool::Instance* conn_pool = cluster_manager_.tcpConnPoolForCluster(
       cluster_name, Upstream::ResourcePriority::Default, this);
   if (!conn_pool) {
+    stats_.no_healthy_upstream_.inc();
```

Comment:
I don't know that this needs to be corrected, but I think service unavailable might be a better name for this stat (and probably the reply message).

The reason is that tcpConnPoolForCluster will return nullptr if the cluster isn't defined which can happen if LDS and CDS don't agree. It will also return nullptr if the there are no healthy hosts in a cluster (and will have incremented the cluster's no healthy hosts counter).

The reply mimics the HTTP router, which also reports "no healthy upstream." The HTTP router, however, charges the request to status code 503 stat, which is more generic than "no healthy upstream".

Reply:
Rather the cluster's upstream_cx_none_healthy stat.

Reply:
We do check if the cluster exists in cluster manager on line 226, so `tcpConnPoolForCluster()` _will_ find the cluster here:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L783

no?

But I am happy with service unavailable too.

Reply:
If we rename it, lets rename this test case too https://github.com/envoyproxy/envoy/blob/master/test/extensions/filters/network/thrift_proxy/router_test.cc#L480.

Reply:
But before that -- am I right that the check on line 226 clears `tcpConnPoolForCluster()` from not finding the cluster?

Reply:
No, you're right. That checks for the cluster first so tcpConnPoolForCluster should only fail on no healthy hosts.